dht_filters.go

```diff
@@ -0,0 +1,137 @@
+package dht
+
+import (
+	"bytes"
+	"net"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	netroute "github.com/libp2p/go-netroute"
+
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+// QueryFilterFunc is a filter applied when considering peers to dial when querying
+type QueryFilterFunc func(dht *IpfsDHT, ai peer.AddrInfo) bool
+
+// RouteTableFilterFunc is a filter applied when considering connections to keep in
+// the local route table.
+type RouteTableFilterFunc func(dht *IpfsDHT, conns []network.Conn) bool
+
+// PublicQueryFilter returns true if the peer is suspected of being publicly accessible
+func PublicQueryFilter(_ *IpfsDHT, ai peer.AddrInfo) bool {
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPublicAddr bool
+	for _, a := range ai.Addrs {
+		if isRelayAddr(a) {
+			return false
+		}
+		if manet.IsPublicAddr(a) {
+			hasPublicAddr = true
+		}
+	}
+	return hasPublicAddr
+}
+
+var _ QueryFilterFunc = PublicQueryFilter
+
+// PublicRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a public network
+func PublicRoutingTableFilter(dht *IpfsDHT, conns []network.Conn) bool {
+	if len(conns) == 0 {
+		return false
+	}
+
+	// Is one of these connections to a public address?
+	for _, c := range conns {
+		addr := c.RemoteMultiaddr()
+		if !isRelayAddr(addr) && manet.IsPublicAddr(addr) {
+			return true
+		}
+	}
+
+	// Do we have a public address for this peer?
+	id := conns[0].RemotePeer()
+	known := dht.peerstore.PeerInfo(id)
+	for _, a := range known.Addrs {
+		if !isRelayAddr(a) && manet.IsPublicAddr(a) {
+			return true
+		}
+	}
+
+	return false
+}
+
+var _ RouteTableFilterFunc = PublicRoutingTableFilter
+
+// PrivateQueryFilter doens't currently restrict which peers we are willing to query from the local DHT.
+func PrivateQueryFilter(dht *IpfsDHT, ai peer.AddrInfo) bool {
+	return len(ai.Addrs) > 0
+}
+
+var _ QueryFilterFunc = PrivateQueryFilter
+
+// PrivateRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a private network
+func PrivateRoutingTableFilter(dht *IpfsDHT, conns []network.Conn) bool {
+	router, _ := netroute.New()
```

Comment:
Is there any way to reuse this? Constructing this doesn't look free.

Reply:
Looking at the implementation, we'd probably need some kind of cache. Let's punt and profile.